File: src/examples/python/binary-tree.py
"""
Binary Tree Traversal - Python

Inorder, preorder, and postorder traversal.
Time Complexity: O(n)
Space Complexity: O(h) where h is tree height
"""

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def inorder(node):
    if node:
        inorder(node.left)
        print(node.val, end=" ")
        inorder(node.right)

def preorder(node):
    if node:
        print(node.val, end=" ")
        preorder(node.left)
        preorder(node.right)

def postorder(node):
    if node:
        postorder(node.left)
        postorder(node.right)
        print(node.val, end=" ")

# Build tree:      1
#                /   \
#               2     3
#              / \
#             4   5
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

print("Inorder:  ", end=""); inorder(root)
print("\nPreorder: ", end=""); preorder(root)
print("\nPostorder:", end=""); postorder(root)

File: src/examples/javascript/binary-tree.js
/**
 * Binary Tree - JavaScript
 *
 * Binary Search Tree with insert and in-order traversal.
 *
 * Time Complexity: O(log n) average for insert/search
 * Space Complexity: O(n)
 */
export default {
  id: 'binary-tree-js',
  language: 'javascript',
  title: 'Binary Search Tree',
  description: 'BST with insert, search, and in-order traversal',
  difficulty: 'intermediate',
  category: 'dataStructures',
  tags: ['data-structures', 'tree', 'recursion', 'BST'],
  code: `class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BST {
  constructor() {
    this.root = null;
  }

  insert(value) {
    const node = new TreeNode(value);
    if (!this.root) {
      this.root = node;
      return;
    }
    let current = this.root;
    while (true) {
      if (value < current.value) {
        if (!current.left) { current.left = node; break; }
        current = current.left;
      } else {
        if (!current.right) { current.right = node; break; }
        current = current.right;
      }
    }
  }

  // In-order: left → root → right (produces sorted output)
  inOrder(node = this.root, result = []) {
    if (!node) return result;
    this.inOrder(node.left, result);
    result.push(node.value);
    this.inOrder(node.right, result);
    return result;
  }
}

const tree = new BST();
[5, 3, 7, 1, 4, 6, 8].forEach(n => tree.insert(n));
console.log("In-order:", tree.inOrder());`,
};
